url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/fr/inria/optimization/cmaes/CMAEvolutionStrategy.java

```diff
@@ -0,0 +1,2704 @@
+    					|| countCupdatesSinceEigenupdate < 1. / sp.getCcov() / N / 5.)) 
+    		return;
+
+    	if (flgdiag) {
+    		for (i = 0; i < N; ++i) {
+    			diagD[i] = Math.sqrt(C[i][i]);
+    		}
+    		countCupdatesSinceEigenupdate = 0;
+        	timings.starteigen = System.currentTimeMillis(); // reset starting time
+        	timings.eigendecomposition = 0;             // not really necessary
+    	} else {
+    		// set B <- C
+    		for (i = 0; i < N; ++i)
+    			for (j = 0; j <= i; ++j)
+    				B[i][j] = B[j][i] = C[i][j];
+
+    		// eigendecomposition
+    		double [] offdiag = new double[N];
+    		long firsttime = System.currentTimeMillis();
+    		tred2(N, B, diagD, offdiag);
+    		tql2(N, diagD, offdiag, B);
+    		timings.eigendecomposition += System.currentTimeMillis() - firsttime;
+
+    		if (options.checkEigenSystem > 0)
+    			checkEigenSystem( N,  C, diagD, B); // for debugging 
+
+    		// assign diagD to eigenvalue square roots
+    		for (i = 0; i < N; ++i) {
+    			if (diagD[i] < 0) // numerical problem?
+    				error("an eigenvalue has become negative");
+    			diagD[i] = Math.sqrt(diagD[i]);
+    		}
+    		countCupdatesSinceEigenupdate = 0;
+    	} // end Update B and D
+    	if (math.min(diagD) == 0) // error management is done elsewhere
+    		axisratio = Double.POSITIVE_INFINITY;
+    	else
+    		axisratio = math.max(diagD) / math.min(diagD);
+
+    } // eigendecomposition
+
+
+    /* ========================================================= */
+    int
+    checkEigenSystem( int N,  double C[][], double diag[], double Q[][]) 
+    /* 
+       exhaustive test of the output of the eigendecomposition
+       needs O(n^3) operations 
+
+       produces error  
+       returns number of detected inaccuracies 
+    */
+    {
+        /* compute Q diag Q^T and Q Q^T to check */
+      int i, j, k, res = 0;
+      double cc, dd; 
+      String s;
+
+      for (i=0; i < N; ++i)
+    	  for (j=0; j < N; ++j) {
+    		  for (cc=0.,dd=0., k=0; k < N; ++k) {
+    			  cc += diag[k] * Q[i][k] * Q[j][k];
+    			  dd += Q[i][k] * Q[j][k];
+    		  }
+    		  /* check here, is the normalization the right one? */
+    		  if (Math.abs(cc - C[i>j?i:j][i>j?j:i])/Math.sqrt(C[i][i]*C[j][j]) > 1e-10 
+    				  && Math.abs(cc - C[i>j?i:j][i>j?j:i]) > 1e-9) { /* quite large */
+    			  s = " " + i + " " + j + " " + cc + " " + C[i>j?i:j][i>j?j:i] + " " + (cc-C[i>j?i:j][i>j?j:i]);
+    			  warning("cmaes_t:Eigen(): imprecise result detected " + s);
+    			  ++res; 
+    		  }
+    		  if (Math.abs(dd - (i==j?1:0)) > 1e-10) {
+    			  s = i + " " + j + " " + dd;
+    			  warning("cmaes_t:Eigen(): imprecise result detected (Q not orthog.) " + s);
+    			  ++res;
+    		  }
+    	  }
+      return res; 
+    }
+
+
+    
+    // Symmetric Householder reduction to tridiagonal form, taken from JAMA package.
+
+    private void tred2 (int n, double V[][], double d[], double e[]) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 23. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1249-1270 into a separate method.